tool/src/main/java/org/datacommons/tool/Processor.java

```diff
@@ -142,4 +229,79 @@ public void checkAllNodes() throws IOException, InterruptedException, DCTooManyF
     }
     existenceChecker.drainRemoteCalls();
   }
+
+  // Called only when resolution is enabled.
+  private void resolveNodes() throws IOException {
+    resolveCommon(
+        McfUtil.mergeGraphs(nodesForVariousChecks),
+        OutputFileType.NODES,
+        OutputFileType.FAILED_NODES);
+  }
+
+  private void resolveCommon(
+      Mcf.McfGraph mcfGraph, OutputFileType successFile, OutputFileType failureFile)
+      throws IOException {
+    McfResolver resolver = new McfResolver(mcfGraph, verbose, idResolver, logCtx);
+    resolver.resolve();
+    if (outputFiles != null) {
+      var resolved = resolver.resolvedGraph();
+      if (!resolved.getNodesMap().isEmpty()) {
+        writeGraph(successFile, resolved);
+      }
+      var failed = resolver.failedGraph();
+      if (!failed.getNodesMap().isEmpty()) {
+        writeGraph(failureFile, failed);
+      }
+    }
+  }
+
+  // Process all the CSV tables to load all external IDs.
+  private void lookupExternalIds() throws IOException, InterruptedException {
+    LogWrapper dummyLog = new LogWrapper(Debug.Log.newBuilder(), Path.of("."));
+    for (var g : nodesForVariousChecks) {
+      for (var idAndNode : g.getNodesMap().entrySet()) {
+        idResolver.submitNode(idAndNode.getValue());
+      }
+    }
+    for (File csvFile : fileGroup.getCsvs()) {
+      if (verbose) logger.info("Reading external IDs from CSV " + csvFile.getPath());
+      TmcfCsvParser parser =
+          TmcfCsvParser.init(
+              fileGroup.getTmcf().getPath(), csvFile.getPath(), fileGroup.delimiter(), dummyLog);
+      if (parser == null) continue;
+      Mcf.McfGraph g;
+      while ((g = parser.parseNextRow()) != null) {
+        for (var idAndNode : g.getNodesMap().entrySet()) {
+          idResolver.submitNode(idAndNode.getValue());
+        }
+      }
+    }
+    idResolver.drainRemoteCalls();
+  }
+
+  private void writeGraph(OutputFileType type, Mcf.McfGraph graph) throws IOException {
+    var writer = writers.getOrDefault(type, null);
+    if (writer == null) {
+      var fileString = outputFiles.get(type).toString();
+      logger.info("Writing to file-type " + type.name() + " in " + fileString);
```

Comment:
nit: Maybe `"Writing to file " + fileString + " of type " + type.name()`

Reply:
Done